qualtran/resource_counting/_qubit_counts.py

```diff
@@ -0,0 +1,132 @@
+                during_size = _bloq_max_width(binst.bloq) + sum(s.shape for s in in_play)
+                max_width = smax(max_width, during_size)
+
+            # After the binst, its successor connections are 'in play'.
+            in_play.update(succ_cxns)
+            after_size = sum(s.shape for s in in_play)
+            max_width = smax(max_width, after_size)
+
+    return max_width
+
+
+@frozen
+class QubitCount(CostKey[int]):
+    """A cost estimating the number of qubits required to implement a bloq.
+
+    The number of qubits is bounded from below by the number of qubits implied by the signature.
+    If a bloq has no callees, the size implied by the signature will be returned. Otherwise,
+    this CostKey will try to compute the number of qubits by inspecting the decomposition.
+
+    In the decomposition, each (sub)bloq is considered to be executed sequentially. The "width"
+    of the circuit (i.e. the number of qubits) at each sequence point is the number of qubits
+    required by the subbloq (computed recursively) plus any "bystander" idling wires.
+
+    This is an estimate for the number of qubits required by an algorithm. Specifically:
+     - Bloqs are assumed to be executed sequentially, minimizing the number of qubits potentially
+       at the expense of greater circuit depth or execution time.
+     - We do not consider "tetris-ing" subbloqs. In a decomposition, each subbloq is assumed
+       to be using all of its qubits for the duration of its execution. This could potentially
+       overestimate the total number of qubits.
+
+    This Min-Max style estimate can provide a good balance between accuracy and scalability
+    of the accounting. To fully account for each qubit and manage space-vs-time trade-offs,
+    you must comprehensively decompose your algorithm to a `cirq.Circuit` of basic gates and
+    use a `cirq.QubitManager` to manage trade-offs. This may be computationally expensive for
+    large algorithms.
+    """
+
+    def compute(self, bloq: 'Bloq', get_callee_cost: Callable[['Bloq'], int]) -> int:
+        """Compute an estimate of the number of qubits used by `bloq`.
+
+        See the class docstring for more information.
+        """
+        # Base case: No callees; use the signature
+        min_bloq_size = bloq.signature.n_qubits()
+        callees = get_bloq_callee_counts(bloq)
+        if len(callees) == 0:
+            logger.info("Computing %s for %s from signature", self, bloq)
+            return min_bloq_size
+
+        # Compute the number of qubits ("width") from the bloq's decomposition. We forward
+        # the `get_callee_cost` function so this can recurse into subbloqs.
+        try:
+            cbloq = bloq.decompose_bloq()
+            logger.info("Computing %s for %s from its decomposition", self, bloq)
+            return _cbloq_max_width(cbloq._binst_graph, get_callee_cost)
+        except (DecomposeNotImplementedError, DecomposeTypeError):
+            pass
```

Comment:
This is a good example of a cost which doesn't scale for large bloqs, as I had raised a concern for in your previous PR (cc https://github.com/quantumlib/Qualtran/issues/957)

Do we have a plan to make this scalable? Do we plan to do `isinstance` based dispatch for special bloqs as you had proposed in https://github.com/quantumlib/Qualtran/pull/913#discussion_r1601989838 ? 

<img width="735" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/2c1ebb9c-b004-4f03-b228-3ab98d5a783e">

Reply:
Here is another example for the hubbard model - 

<img width="507" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/08d05277-f2b7-4dca-b9b9-4702da2d42dd">

Reply:
for something like Add or QROM (which is the slow part in hubbard model) you can annotate the static qubit counts statically

Reply:
I agree. And I think the right way to annotate these static costs is to express the cost as a formula of the cost of its callees. For example - the `SelectSwapQROM` can express its qubit counts in terms of `qubit_count(QROM)` and `qubit_count(SwapWithZero)` bloqs. Similarly, phase estimation can express its cost in terms of `qubit_count(U)`; instead of specifying a constant (which is not even possible for phase estimation since `U` is unknown and decomposing phase estimation can be costly in general, since you end up with a circuit where the depth is at least `2 ** size(phase_register)`)


For this to scale, it would be really nice if `my_static_costs` has access to the cache and can forward it to the `get_cost_cache` method to get the cost of its callees. Since we are designing the API right now, it'll be easier to make this change and be future proof so I'd suggest you to reconsider this proposal.